mono/utils/mono-time.c

```diff
@@ -126,16 +139,57 @@ get_boot_time (void)
 }
 
 /* Returns the number of milliseconds from boot time: this should be monotonic */
+/* Adapted from CoreCLR: https://github.com/dotnet/coreclr/blob/66d2738ea96fcce753dec1370e79a0c78f7b6adb/src/pal/src/misc/time.cpp */
 gint64
 mono_msec_boottime (void)
 {
-	static gint64 boot_time = 0;
-	gint64 now;
-	if (!boot_time)
-		boot_time = get_boot_time ();
-	now = mono_100ns_datetime ();
-	/*printf ("now: %llu (boot: %llu) ticks: %llu\n", (gint64)now, (gint64)boot_time, (gint64)(now - boot_time));*/
-	return (now - boot_time)/10000;
+    gint64 retval = 0;
+
+#if HAVE_CLOCK_MONOTONIC_COARSE || HAVE_CLOCK_MONOTONIC
+    clockid_t clockType =
+#if HAVE_CLOCK_MONOTONIC_COARSE
+        CLOCK_MONOTONIC_COARSE; /* good enough resolution, fastest speed */
+#else
+        CLOCK_MONOTONIC;
+#endif
+    struct timespec ts;
+    if (clock_gettime (clockType, &ts) != 0) {
+        g_error ("clock_gettime(CLOCK_MONOTONIC*) failed; errno is %d", errno, strerror (errno));
+        goto exit;
+    }
+    retval = (ts.tv_sec * tccSecondsToMillieSeconds)+(ts.tv_nsec / tccMillieSecondsToNanoSeconds);
+
+#elif HAVE_MACH_ABSOLUTE_TIME
+    /* use denom == 0 to indicate that s_TimebaseInfo is uninitialised. */
+    if (s_TimebaseInfo.denom == 0) {
+        g_error ("s_TimebaseInfo is uninitialized.");
+        goto exit;
+    }
+    retval = (mach_absolute_time () * s_TimebaseInfo.numer / s_TimebaseInfo.denom) / tccMillieSecondsToNanoSeconds;
+
+#elif HAVE_GETHRTIME
+    retval = (gint64)(gethrtime () / tccMillieSecondsToNanoSeconds);
+
+#elif HAVE_READ_REAL_TIME
+    timebasestruct_t tb;
+    read_real_time (&tb, TIMEBASE_SZ);
+    if (time_base_to_time (&tb, TIMEBASE_SZ) != 0) {
+        g_error ("time_base_to_time() failed; errno is %d (%s)", errno, strerror (errno));
+        goto exit;
+    }
+    retval = (tb.tb_high * tccSecondsToMillieSeconds)+(tb.tb_low / tccMillieSecondsToNanoSeconds);
+
+#else
+    struct timeval tv;
```

Comment:
What do you get in error path? Maybe zero or -1-fll it at entry or in exit path?